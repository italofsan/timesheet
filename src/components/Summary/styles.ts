import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { colors } from "../../styles/colors";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    workedHoursSummaryContainer: {
      display: "flex",
      width: "100%",
      flexDirection: "column",
      alignItems: "flex-start",
      backgroundColor: colors.main,
      padding: "16px 16px",
      borderRadius: 20,
      position: "relative",
    },
    workedHoursSummaryTitle: {
      color: "#eee",
      fontSize: "1.4rem",
    },
    workedHoursSummaryValue: {
      color: "#eee",
      fontSize: "2rem",
    },
    summaryContainer: {
      display: "flex",
      width: "100%",
      flexDirection: "column",
      alignItems: "flex-start",
      backgroundColor: "#fff",
      padding: "16px 16px",
      borderRadius: 20,
      position: "relative",
    },
    summaryTitle: {
      color: "#222",
      fontSize: "1.4rem",
    },

    summaryValue: {
      color: "#222",
      fontSize: "2rem",
    },
    overtimeHoursSummaryValue: {
      color: "#69f0ae",
      fontSize: "2rem",
    },
    undertimeHoursSummaryValue: {
      color: "#ff5252",
      fontSize: "2rem",
    },

    arrowIcon: {
      position: "absolute",
      color: "#ccc",
      bottom: 10,
      right: 10,
      opacity: 0.3,

      fontSize: 96,
      [theme.breakpoints.down("sm")]: {
        fontSize: 48,
      },
    },
  })
);
