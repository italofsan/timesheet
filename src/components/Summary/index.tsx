import { Grid, Paper, Typography } from "@material-ui/core";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import ArrowDownwardIcon from "@material-ui/icons/ArrowDownward";
import HistoryIcon from "@material-ui/icons/History";

import { useTimesheet } from "../../hooks/useTimesheet";

import { subtractHours } from "../../utils/functions/mathHoursFunctions";

import { useStyles } from "./styles";

export function Summary() {
  const classes = useStyles();
  const { totalWorkedHours, totalOvertimeHours, totalUndertimeHours } =
    useTimesheet();

  return (
    <>
      <Grid item xs={12} lg={4}>
        <Paper className={classes.workedHoursSummaryContainer}>
          <Typography className={classes.workedHoursSummaryValue}>
            {totalWorkedHours}
          </Typography>
          <Typography className={classes.workedHoursSummaryTitle}>
            Total Worked Hours
          </Typography>
          <HistoryIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
      <Grid item xs={6} lg={4}>
        <Paper className={classes.summaryContainer}>
          <Typography className={classes.overtimeHoursSummaryValue}>
            {subtractHours(totalUndertimeHours, totalOvertimeHours).length === 6
              ? "00:00"
              : `${subtractHours(totalUndertimeHours, totalOvertimeHours)}`}
          </Typography>
          <Typography className={classes.summaryTitle}>
            Total Overtime Hours
          </Typography>
          <ArrowUpwardIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
      <Grid item xs={6} lg={4}>
        <Paper className={classes.summaryContainer}>
          <Typography className={classes.undertimeHoursSummaryValue}>
            {subtractHours(totalOvertimeHours, totalUndertimeHours).length >= 6
              ? "00:00"
              : `${subtractHours(totalOvertimeHours, totalUndertimeHours)}`}
          </Typography>
          <Typography className={classes.summaryTitle}>
            Total Undertime Hours
          </Typography>
          <ArrowDownwardIcon className={classes.arrowIcon} />
        </Paper>
      </Grid>
    </>
  );
}
