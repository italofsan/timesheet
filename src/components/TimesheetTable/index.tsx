import { useState } from "react";
import {
  Grid,
  IconButton,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Tooltip,
  Zoom,
} from "@material-ui/core";
import { Edit, Delete } from "@material-ui/icons";

import { UpdateModal } from "../../components/Modals/UpdateModal";
import { useTimesheet } from "../../hooks/useTimesheet";

import { DataTimesheet } from "../../utils/types";

import { useStyles } from "./styles";
import { DeleteModal } from "../Modals/DeleteModal";
import { colors } from "../../styles/colors";

const columnsTimesheetTable = [
  "Options",
  "Date",
  "Arrival",
  "Lunch Exit",
  "Lunch Arrival",
  "Exit",
  "Worked Hours",
  "Overtime",
  "Undertime",
];

export function TimesheetTable() {
  const classes = useStyles();
  const { listTimesheets } = useTimesheet();

  const [selectedTimesheet, setSelectedTimesheet] = useState<DataTimesheet>(
    {} as DataTimesheet
  );

  const [updateModalOpen, setUpdateModalOpen] = useState(false);
  function handleOpenUpdateModal() {
    setUpdateModalOpen(true);
  }
  function handleCloseUpdateModal() {
    setUpdateModalOpen(false);
  }
  const [deleteModalOpen, setDeleteModalOpen] = useState(false);
  function handleOpenDeleteModal() {
    setDeleteModalOpen(true);
  }
  function handleCloseDeleteModal() {
    setDeleteModalOpen(false);
  }

  return (
    <Grid item xs={12}>
      <TableContainer component={Paper}>
        <Table>
          <caption
            style={{ backgroundColor: colors.main, color: "#FFF" }}
          >{`You have ${
            listTimesheets && listTimesheets.length !== 0
              ? listTimesheets.length
              : 0
          }  register${
            listTimesheets && listTimesheets.length !== 1 ? "s" : ""
          }!`}</caption>
          <TableHead>
            <TableRow>
              {columnsTimesheetTable.map((columnTimesheetTable, index) => (
                <TableCell
                  key={index}
                  align='center'
                  className={classes.tableCellHeadOdd}
                >
                  {columnTimesheetTable}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {listTimesheets.map((timesheet, index) => (
              <TableRow key={index}>
                <TableCell align='center'>
                  <Tooltip
                    arrow
                    title='Edit'
                    placement='top'
                    enterDelay={500}
                    TransitionComponent={Zoom}
                  >
                    <IconButton
                      edge='start'
                      className={classes.optionsButton}
                      color='inherit'
                      aria-label='edit'
                      onClick={() => {
                        setSelectedTimesheet(timesheet);
                        handleOpenUpdateModal();
                      }}
                    >
                      <Edit />
                    </IconButton>
                  </Tooltip>
                  <Tooltip
                    arrow
                    title='Delete'
                    placement='top'
                    enterDelay={500}
                    TransitionComponent={Zoom}
                  >
                    <IconButton
                      edge='start'
                      className={classes.optionsButton}
                      color='inherit'
                      aria-label='delete'
                      onClick={() => {
                        setSelectedTimesheet(timesheet);
                        handleOpenDeleteModal();
                      }}
                    >
                      <Delete />
                    </IconButton>
                  </Tooltip>
                </TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.date &&
                    new Intl.DateTimeFormat("en-US").format(
                      new Date(timesheet.date)
                    )}
                </TableCell>
                <TableCell align='center'>{timesheet.arrivalTime}</TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.lunchExitTime}
                </TableCell>
                <TableCell align='center'>
                  {timesheet.lunchArrivalTime}
                </TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.exitTime}
                </TableCell>
                <TableCell align='center'>{timesheet.workedHours}</TableCell>
                <TableCell align='center' style={{ backgroundColor: "#eef" }}>
                  {timesheet.overtimeHours}
                </TableCell>
                <TableCell align='center'>{timesheet.undertimeHours}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      <UpdateModal
        selectedTimesheet={selectedTimesheet}
        updateModalOpen={updateModalOpen}
        handleCloseUpdateModal={handleCloseUpdateModal}
      />
      <DeleteModal
        selectedTimesheet={selectedTimesheet}
        deleteModalOpen={deleteModalOpen}
        handleCloseDeleteModal={handleCloseDeleteModal}
      />
    </Grid>
  );
}
