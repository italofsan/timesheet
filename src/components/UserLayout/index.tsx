import { ReactNode } from "react";
import {
  AppBar,
  Button,
  Container,
  Grid,
  Toolbar,
  Typography,
} from "@material-ui/core";

import { useHistory } from "react-router-dom";

import { useStyles } from "./styles";
import { useAuth } from "../../hooks/useAuth";

interface UserLayoutProps {
  children: ReactNode;
}

const UserLayout = ({ children }: UserLayoutProps) => {
  const classes = useStyles();
  const history = useHistory();
  const { signOut } = useAuth();

  function handleSignOut() {
    signOut();
    history.push("/");
  }

  return (
    <div className={classes.root}>
      <AppBar position='fixed' className={classes.appBar}>
        <Container maxWidth='lg'>
          <Toolbar>
            <Typography variant='h6' className={classes.title}>
              Timesheet App
            </Typography>
            <Button color='inherit' onClick={handleSignOut}>
              Logout
            </Button>
          </Toolbar>
        </Container>
      </AppBar>
      <Grid item xs={12}>
        <main className={classes.content}>
          <Toolbar />
          {children}
        </main>
      </Grid>
    </div>
  );
};

export default UserLayout;
