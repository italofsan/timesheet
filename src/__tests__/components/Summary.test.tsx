import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import { Summary } from "../../components/Summary";

describe("Summary component", () => {
  it("renders correctly", () => {
    render(<Summary />);

    expect(screen.getByText("Total Worked Hours")).toBeInTheDocument();
    expect(screen.getByText("Total Overtime Hours")).toBeInTheDocument();
    expect(screen.getByText("Total Undertime Hours")).toBeInTheDocument();
  });

  const tree = renderer.create(<Summary />).toJSON();
  expect(tree).toMatchSnapshot();
});
