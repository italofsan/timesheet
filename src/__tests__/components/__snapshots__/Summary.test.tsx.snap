// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-4"
  >
    <div
      className="MuiPaper-root makeStyles-workedHoursSummaryContainer-1 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        className="MuiTypography-root makeStyles-workedHoursSummaryValue-3 MuiTypography-body1"
      />
      <p
        className="MuiTypography-root makeStyles-workedHoursSummaryTitle-2 MuiTypography-body1"
      >
        Total Worked Hours
      </p>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-arrowIcon-9"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-lg-4"
  >
    <div
      className="MuiPaper-root makeStyles-summaryContainer-4 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        className="MuiTypography-root makeStyles-overtimeHoursSummaryValue-7 MuiTypography-body1"
      >
        00:00
      </p>
      <p
        className="MuiTypography-root makeStyles-summaryTitle-5 MuiTypography-body1"
      >
        Total Overtime Hours
      </p>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-arrowIcon-9"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
        />
      </svg>
    </div>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-lg-4"
  >
    <div
      className="MuiPaper-root makeStyles-summaryContainer-4 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        className="MuiTypography-root makeStyles-undertimeHoursSummaryValue-8 MuiTypography-body1"
      >
        00:00
      </p>
      <p
        className="MuiTypography-root makeStyles-summaryTitle-5 MuiTypography-body1"
      >
        Total Undertime Hours
      </p>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-arrowIcon-9"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
        />
      </svg>
    </div>
  </div>,
]
`;
